lib/spack/llnl/util/filesystem.py

```diff
@@ -1281,6 +1284,17 @@ def traverse_tree(
         dest_child = os.path.join(dest_path, f)
         rel_child = os.path.join(rel_path, f)
 
+        # If the source path is a link and the link's source is ignored, then ignore the link too.
+        if os.path.islink(source_child) and not follow_links:
+            target = os.readlink(source_child)
+            all_parents = accumulate(target.split(os.sep), lambda x, y: os.path.join(x, y))
+            if any(map(ignore, all_parents)):
+                tty.warn(
+                    f"Skipping {source_path} because the source or a part of the source's "
+                    f"path is included in the ignores."
+                )
+                continue
```

Comment:
I would disagree that the `ignore` function is only files, as the `filesystem.py::TestCopyTree::test_symlinks_true_ignore` unit test ignores `"a"`, which is a directory. If `ignore` is only supposed to be files, then we should check for that and raise an exception if someone tries to ignore a directory.

Assuming that using directories in the `ignore` function is okay:
This isn't exactly excluding symlinks, it's simply not creating a symlink if the link would be broken because the target or its parent directory is in `ignore`. What is the use case that a user would want to ignore the link's source but keep the (now broken) link?

Reply:
> I would disagree that the ignore function is only files,

I was wrong about that - thanks for the pointer!

> This isn't exactly excluding symlinks, it's simply not creating a symlink if the link would be broken because the target or its parent directory is in ignore. 

Good point!

Although this may need some more care WRT relative symlinks: `resolve_link_target_relative_to_the_link` could be useful but resolves to an absolute path, and I'm not sure if that will interfere with `ignore` (perhaps "re-relative-izing it WRT the source root would work).

This also brings up a broader point about the robustness of traversals: depending on the traversal order you may see symlinks to files before the files themselves (at which point, you can't know whether that file will or won't exist, and likewise whether the symlink will dangle): I wonder if replacing this recursive/yield approach with something that computes everything up front would help with that.

Reply:
Is this what you mean by "re-relative-izing" with relation to the source root? This is the code in `traverse_tree` with a few extra lines added to de-then-re-relativize the link target (source)

```python
>>> source_path = '/filesystem_test0/source'
>>> source_child = '/filesystem_test0/source/f'
>>> target = os.readlink(source_child)
>>> print(target)
a/b
>>> resolved_link_target = resolve_link_target_relative_to_the_link(source_child)
>>> print(resolved_link_target)
/filesystem_test0/source/a/b
>>> rerelativized_link = os.path.relpath(resolved_link_target, source_path)
>>> print(rerelativized_link)
a/b
>>> all_parents = accumulate(rerelativized_link.split(os.sep), lambda x, y: os.path.join(x, y))
>>> print(all_parents)
['a', 'a/b']
>>> ignore = lambda x: x in ['c/d/e', 'a']
>>> any(map(ignore, all_parents))
True
```

It seems circular to me, but maybe this example is too simple.

> This also brings up a broader point about the robustness of traversals: depending on the traversal order you may see symlinks to files before the files themselves (at which point, you can't know whether that file will or won't exist, and likewise whether the symlink will dangle): I wonder if replacing this recursive/yield approach with something that computes everything up front would help with that.

When we find a symlink, it checks the output of `os.readlink()` and looks if _that_ path or its parent directories are ignored. If that path is in the ignores, the link is skipped, too. It shouldn't matter what order they are found in. I don't think that it's necessary to build the whole directory tree first to know that info.